challenge_serving_drinks/src/challenge_serving_drinks/sd_states.py

```diff
@@ -0,0 +1,539 @@
+
+                    return "succeeded"
+
+            else:
+                # We should not end up here
+                rospy.logerr("Missing drinks information")
+
+                # Try to get the answer
+                try:
+                    speech_result = self._query_drink(self._max_queries_per_try)
+                except TimeoutException:
+                    continue
+
+                # Ask for confirmation
+                self._robot.speech.speak(
+                    "I understood that you would like {}, is that correct?".format(speech_result.semantics)
+                )
+                if not self._confirm():
+                    continue
+
+                # Store the type in the designator
+                self._drink_designator.write(str(speech_result.semantics))
+                rospy.loginfo("Drink to fetch: {}".format(self._drink_designator.resolve()))
+
+                return "succeeded"
+
+        self._robot.speech.speak("I am sorry but I cannot understand you. I will quit now", block=False)
+        # ToDo: fill default?
+        self._robot.head.cancel_goal()
+        return "failed"
+
+    def _query_drink(self, max_tries):
+        """
+        Queries the HMI server for <max_tries> times until a result has been obtained
+        :return: (HMIResult) (N.B.: the 'semantics' attribute should be a dict with 'beverage' key)
+        :raises: (TimeoutException) if robot didn't hear anything
+        """
+        attempt = 0
+        while not rospy.is_shutdown():
+            attempt += 1
+            try:
+                speech_result = self._robot.hmi.query(description="Can I please take your order",
+                                                      grammar=self._drinks_grammar,
+                                                      target=self._drinks_target)
+                return speech_result
+            except TimeoutException:
+                if attempt >= max_tries:
+                    raise  # Re-raise
+
+            # We only end up here if the exception was caught
+            self._robot.speech.speak(random.choice([
+                "I'm sorry, can you repeat",
+                "Please repeat your order, I didn't hear you",
+                "I didn't get your order, can you repeat it",
+                "Please speak up, as I didn't hear your order",
+            ]))
+
+    def _confirm(self):
+        """
+        Queries the HMI server for confirmation
+        :return: (bool) whether the understood beverage was correct
+        """
+        try:
+            speech_result = self._robot.hmi.query(description="Is this correct?", grammar="T[True] -> yes;"
+                                                                                          "T[False] -> no", target="T")
+        except TimeoutException:
+            return False
+
+        return speech_result.semantics
+
+
+class AskAvailability(smach.State):
```

Comment:
Samwe, this could be a SM which I think would make debugging easier.

Reply:
Not completely necessary, as discussed. Will be cleaned a bit more in the new PR